src/query/xasl_cache.c

```diff
@@ -1163,6 +1176,86 @@ xcache_entry_mark_deleted (THREAD_ENTRY * thread_p, XASL_CACHE_ENTRY * xcache_en
   return (new_cache_flag == XCACHE_ENTRY_DELETED_BY_ME);
 }
 
+/*
+* xcache_entry_set_request_recompile_flag () - Mark XASL cache entry as "request recompile".
+*
+* return	    : True if the flag was successfuly set (or cleared).
+* thread_p (in)     : Thread entry.
+* xcache_entry (in) : XASL cache entry.
+* set_flag(in)      : true if flag should be set, false if should be cleared
+*/
+static bool
+xcache_entry_set_request_recompile_flag (THREAD_ENTRY * thread_p, XASL_CACHE_ENTRY * xcache_entry, bool set_flag)
+{
+  INT32 cache_flag = 0;
+  INT32 new_cache_flag;
+
+  /* Mark for delete. We must successfully set XCACHE_ENTRY_MARK_DELETED flag. */
+  do
+    {
+      cache_flag = xcache_entry->xasl_id.cache_flag;
+
+      if (cache_flag & XCACHE_ENTRY_MARK_DELETED)
+	{
+	  xcache_log ("tried to set flag request recompile, but entry is marked for delete: \n"
+		      XCACHE_LOG_ENTRY_TEXT ("entry") XCACHE_LOG_TRAN_TEXT,
+		      XCACHE_LOG_ENTRY_ARGS (xcache_entry), XCACHE_LOG_TRAN_ARGS (thread_p));
+	  return false;
+	}
+
+      if (set_flag && (cache_flag & XCACHE_ENTRY_TO_BE_RECOMPILED))
+	{
+	  /* Somebody is compiling the entry already we are too late */
+	  xcache_log_error ("tried to mark entry as request recompile, but it was marked as to be recompiled: \n"
+			    XCACHE_LOG_ENTRY_TEXT ("entry") XCACHE_LOG_TRAN_TEXT,
+			    XCACHE_LOG_ENTRY_ARGS (xcache_entry), XCACHE_LOG_TRAN_ARGS (thread_p));
+	  return false;
+	}
+
+      if (!set_flag && (cache_flag & XCACHE_ENTRY_TO_BE_RECOMPILED))
```

Comment:
What happens with ```if (set_flag && (cache_flag & XCACHE_ENTRY_WAS_RECOMPILED))``` case? Marking a recompiled entry will not help, because it won't be visible to anyone. I think we have the options to bail or to assert it as unacceptable.